Add a dedicate queue for incoming ClusterStates

bleskes opened the pull request:
The initial implementation of two phase commit based cluster state publishing (#13062) relied on a single in memory "pending" cluster state that is only processed by ZenDiscovery once committed by the master. While this is fine on it's own, it resulted in an issue with acknowledged APIs, such as the open index API, in the extreme case where a node falls behind and receives a commit message after a new cluster state has been published. Specifically:

1) Master receives and acked-API call and publishes cluster state CS1
2) Master waits for a min-master nodes to receives CS1 and commits it.
3) All nodes that have responded to CS1 are sent a commit message, however, node N didn't respond yet
4) Master waits for publish timeout (defaults to 30s) for all nodes to process the commit. Node N fails to do so.
5) Master publishes a cluster state CS2. Node N responds to cluster state CS1's publishing but receives cluster state CS2 before the commit for CS1 arrives.
6) The commit message for cluster CS1 is processed on node N, but fails because CS2 is pending. This caused the acked API in step 1 to return (but CS2 , is not yet processed).

In this case, the action indicated by CS1 is not yet executed on node N and therefore the acked API calls return pre-maturely. Note that once CS2 is processed but the change in CS1 takes effect (cluster state operations are safe to batch and we do so all the time).

An example failure can be found on: http://build-us-00.elastic.co/job/es_feature_two_phase_pub/314/

This commit extracts the already existing pending cluster state queue (`processNewClusterStates`) from ZenDiscovery into it's own class, which serves as a temporary container for in-flight cluster states. Once committed the cluster states are transferred to ZenDiscovery as they used to before. This allows "lagging" cluster states to still be successfully committed and processed (and likely to be ignored as a newer cluster state has already been processed).  
As a side effect, all batching logic is now extracted from ZenDiscovery and is unit tested.

Note: this PR is done against the feature/two-phase-pub branch

bleskes commented:
One more note - this commit changes the beavior of how we respond to an incoming cluster state while the local node is master. Before we would tie break and step down or ask the other master to step down. This is removed due to two reasons:
1) We already have this logic when receiving node fault detection pings from the other master.
2) It is better to reject the incoming cluster state (as would other nodes do) to avoid the other master committed it by mistake if the two masters form together a quorum (3 master nodes scenario).

s1monw commented:
+1 just for the sake of the test

brwe commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java:
```diff
@@ -0,0 +1,254 @@
+        for (; index < pendingStates.size(); index++) {
+            ClusterStateContext potentialState = pendingStates.get(index);
+
+            if (potentialState.state.supersedes(stateToProcess.state) && potentialState.committed()) {
```
We assume in this method hat cluster states are ordered? But can we really assume that? What if cluster states arrive out of order and queue looks like this: c1(committed), c3(not committed), c2 (committed)? c2 will then not be applied until c3 is committed.

bleskes replied:
cluster states can indeed come out of order as far as this class is concerned (see tests). I the example you give cluster c2 will be processed together with c1 (to be exact, instead of it). Note that this loop goes on all the way to the end of the list.

brwe replied:
yeah, all good. I misunderstood.

brwe commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java:
```diff
@@ -0,0 +1,254 @@
+            final DiscoveryNode pendingMasterNode = pendingState.nodes().masterNode();
+            if (Objects.equals(currentMaster, pendingMasterNode) == false) {
+                contextsToRemove.add(pendingContext);
+                if (pendingContext.listener != null) {
```
pendingContext.listener != null -> pendingContext.comitted() ?

bleskes replied:
I originally went with "there is a listener, so we should use it" logic, not caring much about what it means to have a listener. At the time I had a committed boolean as well. When I removed it, I felt it's still clearer this way but sure - can change.

brwe commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -682,171 +670,90 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
-            this.clusterState = clusterState;
-        }
-    }
+    void handleNewClusterStateFromMaster(String reason) {
```
can we call this something different, like applyLatestCommitedClusterState ?

bleskes replied:
I agrees the current name carries less meaning in the way it works now. applyLastCommitedClusterState is not ideal in my mind as we don't know which state will be selected. that's up to the queue. How about - processNextPendingClusterState? I'll use this if there is no objection...

dakrone commented on core/src/main/java/org/elasticsearch/cluster/ClusterState.java:
```diff
@@ -296,6 +293,11 @@ public String toString() {
         }
     }
 
+    public boolean supersedes(ClusterState other) {
```
Please add javadocs for this since it's a public method

dakrone replied:
And even better would be a static method for superseding that could be unit tested

bleskes replied:
I added java docs and a test. I think this is more natural than a static method and doesn't prevent us from testing?

dakrone commented on core/src/main/java/org/elasticsearch/cluster/ClusterState.java:
```diff
@@ -296,6 +293,11 @@ public String toString() {
     }
 
+    public boolean supersedes(ClusterState other) {
+        return this.nodes().masterNodeId() != null && this.nodes().masterNodeId().equals(other.nodes().masterNodeId()) && this.version() > other.version();
```
Can `other.nodes().masterNodeId()` be null? Do we need a check there?

bleskes replied:
string.equals deals with this correctly..

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -634,9 +624,7 @@ public ClusterState execute(ClusterState currentState) {
-                ArrayList<ProcessClusterState> pendingNewClusterStates = new ArrayList<>();
-                processNewClusterStates.drainTo(pendingNewClusterStates);
-                logger.trace("removed [{}] pending cluster states", pendingNewClusterStates.size());
+                publishClusterState.pendingStatesQueue().failAllStatesAndClear(new ElasticsearchException("master left [" + reason + "]"));
```
`ElasticsearchException` supports the `{}` formatting like logging does, using that we can remove the string concatenation (and subsequent extra object creation) here.

bleskes replied:
check

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -682,171 +670,90 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
-                    newStateProcessed.onNewClusterStateProcessed();
-                }
+                assert newClusterState.nodes().masterNode() != null : "received a cluster state without a master";
+                assert !newClusterState.blocks().hasGlobalBlock(discoverySettings.getNoMasterBlock()) : "received a cluster state with a master block";
```
Why should we keep these as asserts instead of hard checks? They seem like they should never happen regardless.

bleskes replied:
These used to be there, so I left them. In my opinion asserts are things we know to be impossible because they are checked in other places in the code (with hard exceptions). They are meant as an extra safety to signal that stuff "smells" and also document what the "expected" state is according to the assumption the code is working on. I don't feel strongly about this though. I can remove them here or convert to exceptions.

dakrone replied:
I think we should convert them to exceptions so we can see if a user ever runs into them in production.

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -682,171 +670,90 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                    for (IndexMetaData indexMetaData : newClusterState.metaData()) {
+                        IndexMetaData currentIndexMetaData = currentState.metaData().index(indexMetaData.index());
+                        if (currentIndexMetaData != null && currentIndexMetaData.isSameUUID(indexMetaData.indexUUID()) &&
+                                currentIndexMetaData.version() == indexMetaData.version()) {
```
I think this check is long enough to warrant a separate (and unit testable!) method for checking if the index metadata is the same

bleskes replied:
yeah, agreed, but it's a different change (this is the same code as before).

dakrone replied:
Okay, want to open an issue so we don't forget?

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java:
```diff
@@ -0,0 +1,254 @@
+/**
+ * A queue that holds all "in-flight" incoming cluster states from the master. Once a master commits a cluster
+ * state, it is made available via {@link #getNextClusterStateToProcess()}. The class also takes care of batching
+ * cluster states for processing and failures
```
Maybe add a note about the thread-safety guarantees of this class?

bleskes replied:
added

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java:
```diff
@@ -0,0 +1,254 @@
+            throw new IllegalStateException("can't resolve failed cluster state with uuid [" + state.stateUUID() + "], version [" + state.version() + "]");
+        }
+        final ClusterStateContext failedContext = pendingStates.get(failedIndex);
+        assert failedContext.committed() : "failed cluster state is not committed " + state;
```
Should this be a hard check?

bleskes replied:
yeah, can do.

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java:
```diff
@@ -0,0 +1,254 @@
+        }
+
+        void markAsCommitted(StateProcessedListener listener) {
+            assert this.listener == null : "double committing of " + state;
```
Should we bail here if it's marked twice and throw an exception?

bleskes replied:
yeah, can do...

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java:
```diff
@@ -0,0 +1,254 @@
+
+        @Override
+        public String toString() {
+            return "[uuid[" + stateUUID() + "], v[" + state.version() + "], m[" + state.nodes().masterNodeId() + "]]";
```
Would it be nicer to use a StringBuilder here?

bleskes replied:
question of taste :) 

To me the above is more readable than:

```
  return new StringBuilder("[uuid[").append(stateUUID()).append("], v[").append(state.version()).append("], m[").append(state.nodes().masterNodeId()).append("]]").toString();

```

dakrone replied:
There's nothing saying you have to put the StringBuilder on a single line, that's the point :)

```java
StringBuilder sb = new StringBuilder("[uuid[");
sb.append(stateUUID());
sb.append("], v[").append(state.version);
sb.append("], m[").append(state.nodes().masterNodeId());
sb.append("]]");
return sb.toString();
```

jasontedor replied:
If it's just for readability, how's this?

    String.format(
        Locale.ROOT,
        "[uuid[%s], v[%d], m[%s]]",
        stateUUID(),
        state.version(),
        state.nodes().masterNodeId()
    )

bleskes replied:
Jason wins :)

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java:
```diff
@@ -382,56 +383,33 @@ void validateIncomingState(ClusterState incomingState, ClusterState lastSeenClus
+            @Override
+            public void onNewClusterStateProcessed() {
+                try {
+                    channel.sendResponse(TransportResponse.Empty.INSTANCE);
```
I know this code is mimicking what the old stuff did, but can you add an explanation for why an empty response is sent?

bleskes replied:
added something.

dakrone commented on core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java:
```diff
@@ -961,6 +963,15 @@ public void testIndicesDeleted() throws Exception {
 
+    @After
+    // make sure we do not leak pending cluster states
+    public void assertNoPendingClusterStates() throws Exception {
```
Out of curiosity, do we want the check on a global level as well as just for this IT?

bleskes replied:
will try.

dakrone commented:
Went through this and left some questions.

I have a naive question, instead of an unbounded queue that could potentially grow very large, can we cap the queue at a specific length, and when a new state comes in, drop the oldest state currently in the queue?

imotov commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java:
```diff
@@ -0,0 +1,254 @@
+
+    }
+
+    int findState(String stateUUID) {
```
Since the actual index itself doesn't seem to be used, maybe just return the state itself if it's found or null if state is not found and hide the whole array abstraction away?

bleskes replied:
I think you mean the context. Good idea. changed.

imotov commented:
Left one minor comment, but I have the same question that Lee asked above about unbounded queue growth.

bleskes commented:
@dakrone @imotov @brwe  I addressed all comments.

Regarding unbound queue- yeah, I was thinking about it but it's tricky (FWIW - the queue was unbound before, but we were also never holding on to things more than until a cluster state thread came along). First question is - how long should it be? second, what should you do when the queue is full? you can reject incoming cluster states from the master, which may or may not prevent them from being committed. But once they are rejected (but committed by the master), the node will never process them. This probably something we should deal with as well (upon what publishing failures should we remove a node from the cluster?) but I rather not address it now. 

I suggest that we flush pending states that have been in the queue for some ridiculous time - say 10m. That will make sure that we never remove states that are "leaked" and will also make sure no state is rejected. Also, it will help with the testClusterJoinDespiteOfPublishingIssues test that leaks cluster states "by design" by swallowing commit requests. Note though that as far as I know that is impossible in real networks (unless someone messes up with the iptables)

If you agree I'll do that in a follow up commit. We also need some monitoring tool for this (another issue). Until then I added an awaitFix to the `testClusterJoinDespiteOfPublishingIssues` test

bleskes commented:
I chatted more with @dakrone and we came up with a simpler solution - bound the queue and just drop old cluster states if need be (see java docs for the reasoning why that's OK to do). I also worked around the expected leakage in testClusterJoinDespiteOfPublishingIssues.

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -682,171 +670,90 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
-        }
-    }
+    void processNextPendingClusterState(String reason) {
+        clusterService.submitStateUpdateTask("zen-disco-receive(from master [" + reason + "])", Priority.URGENT, new ProcessedClusterStateNonMasterUpdateTask() {
```
So, completely separate change from this PR, but how do you feel about renaming all of these log messages from "zen-disco-whatever" to either "zen-whatever" or "zend-whatever"? **Every** time I read "zen-disco" I think "zen-disconnect" instead of "zen-discovery" because they share the same prefix.

bleskes replied:
++ :)

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -682,171 +670,90 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+            public void onFailure(String source, Throwable t) {
+                logger.error("unexpected failure during [{}]", t, source);
+                if (newClusterState != null) {
+                    publishClusterState.pendingStatesQueue().markAsFailed(newClusterState, t);
```
Can marking as failed fail? Should this be wrapped in a try/catch?

bleskes replied:
added logging. Nothing much we can do.

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java:
```diff
@@ -0,0 +1,279 @@
+ *         i) we are no longer following the master of the dropped cluster state but follow the incoming one
+ *         ii) we are no longer following any master, in which case it doesn't matter which cluster state will be processed first.
+ *
+ * The class is fully thread safe and can be used concurrently.
```
Thanks for adding this!

dakrone commented on core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java:
```diff
@@ -57,32 +57,30 @@
+        this.newPendingClusterStatelistener = listener;
         this.discoverySettings = discoverySettings;
         this.clusterName = clusterName;
+        this.pendingStatesQueue = new PendingClusterStatesQueue(logger, settings.getAsInt(SETTINGS_MAX_PENDING_CLUSTER_STATES, 100));
```
I think we should lower this to 25, since it should practically never be higher than 1 in healthy cases. What do you think?

dakrone replied:
I just want to make sure we don't hit an OOME with huuuuge cluster states

bleskes replied:
OK :)

dakrone commented on core/src/test/java/org/elasticsearch/cluster/ClusterStateTests.java:
```diff
@@ -0,0 +1,54 @@
+
+public class ClusterStateTests extends ESTestCase {
+
+    public void testSupersedes() {
```
Hooray for unit tests!

dakrone commented on core/src/test/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueueTests.java:
```diff
@@ -0,0 +1,224 @@
+            assertThat(((MockListener) context.listener).failure, notNullValue());
+        }
+
+        // all states that should have dropped are
```
all states that should have dropped are what?

bleskes replied:
are indeed dropped.

dakrone commented:
@bleskes left some more (mostly minor) comments on this

bleskes commented:
@dakrone thx. Pushed another update.

dakrone commented:
Thanks, LGTM

bleskes commented:
Thx @dakrone @imotov & @brwe . I push this to feature/two_phase_pub to give CI some time.
